File: samples/Test/Classes/RootWindow.cpp


#include "RootWindow.h"
#include "MenuViewController.h"
#include "FirstViewController.h"
#include "MyWebViewController.h"
#if (CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID)
#include <jni.h>
#include "platform/android/jni/JniHelper.h"
#endif

static RootWindow* _window = NULL;

RootWindow* RootWindow::getInstance()
{
    if (_window == NULL)
    {
        _window = new RootWindow();
        _window->init();
        _window->autorelease();
    }
    return _window;
    
}

RootWindow::RootWindow()
:m_pRootNavigationController(NULL)
,m_pRootDrawerController(NULL)
{
    CAApplication::getApplication()->getKeypadDispatcher()->onBackClicked([&]
    {
        if (this->getModalViewController())
        {
            this->dismissModalViewController(true);
        }
        else if (this->getDrawerController()->isShowLeftViewController())
        {
            this->getDrawerController()->hideLeftViewController(true);
        }
        else if (this->getRootNavigationController()->getViewControllerCount() > 1)
        {
            this->getRootNavigationController()->popViewControllerAnimated(true);
        }
        else
        {
            CAApplication::getApplication()->end();
        }
    });
    
    CAApplication::getApplication()->setStatusBarStyle(CAStatusBarStyle::LightContent);
}

RootWindow::~RootWindow()
{
}

bool RootWindow::init()
{
    if (!CAWindow::init())
    {
        return false;
    }
    
    CAVector<CAViewController*> conttollers;
    for (int i=0; i<3; i++)
    {
        conttollers.pushBack(FirstViewController::create());
    }

    CANavigationBarItem* firstItem = CANavigationBarItem::create(UTF8("控件展示"));
    CABarButtonItem* barButtonItem = CABarButtonItem::create("", CAImage::create("image/ic_category_list.png"), NULL);
    barButtonItem->setCallbackFunction([=]()
    {
        this->getDrawerController()->showLeftViewController(true);
    });
    firstItem->addLeftButtonItem(barButtonItem);
    
    CATabBarController* tabBarConttoller = new CATabBarController();
    tabBarConttoller->initWithViewControllers(conttollers);
    tabBarConttoller->setNavigationBarItem(firstItem);
    tabBarConttoller->autorelease();

    m_pRootNavigationController = new CANavigationController();
    m_pRootNavigationController->initWithRootViewController(tabBarConttoller);
    m_pRootNavigationController->autorelease();

    CADrawerController* drawer = new CADrawerController();
    drawer->initWithController(MenuViewController::create(), m_pRootNavigationController);
    drawer->autorelease();
    drawer->setBackgroundImage(CAImage::create("image/bg.jpg"));
    drawer->setEffect3D(true);
    
    this->setRootViewController(drawer);
    
    m_pRootDrawerController = drawer;

    return true;
}

File: samples/Test_JS/Resources/js/rootwindow.js


var RootWindow = ca.CAWindow.extend(
    {
        m_pRootNavigationController:null,
        ctor: function () {
            this._super();

            var viewController = new FirstViewController();

            var controllers = [];
            for(var i=0; i<3; i++)
            {
                controllers.push(new FirstViewController());
            }

            var firstItem = ca.CANavigationBarItem.create("控件展示" + "<" + ca.winSize.width + "x" + ca.winSize.height + ">");
            var barButtonItem = ca.CABarButtonItem.create("", ca.CAImage.create("image/ic_category_list.png"), null);
            barButtonItem.setCallbackFunction(function (){
                this.drawer.showLeftViewController(true);
            }.bind(this));
            firstItem.addLeftButtonItem(barButtonItem);

            this.tabBarConttoller = new ca.CATabBarController();
            this.tabBarConttoller.initWithViewControllers(controllers);
            this.tabBarConttoller.setNavigationBarItem(firstItem);

            this.m_pRootNavigationController = new ca.CANavigationController();
            this.m_pRootNavigationController.initWithRootViewController(this.tabBarConttoller);


            this.drawer = new ca.CADrawerController();
            this.drawer.initWithController(new MenuViewController(), this.m_pRootNavigationController);
            this.drawer.autorelease();
            this.drawer.setBackgroundImage(ca.CAImage.create("image/bg.jpg"));
            this.drawer.setEffect3D(true);

            this.setRootViewController(this.drawer);

        },
        getRootNavigationController:function(){
            return this.m_pRootNavigationController;
        },
        getDrawerController:function () {
            return this.drawer;
        },
    }
);
